src/test/java/software/amazon/awssdk/crt/test/HttpRequestResponseTest.java

```diff
@@ -0,0 +1,318 @@
+            actuallyConnected = true;
+            CrtHttpStreamHandler responseHandler = new CrtHttpStreamHandler() {
+                @Override
+                public void onResponseHeaders(HttpStream stream, int responseStatusCode, HttpHeader[] nextHeaders) {
+                    response.statusCode = responseStatusCode;
+                    response.headers.addAll(Arrays.asList(nextHeaders));
+                }
+
+                @Override
+                public void onResponseHeadersDone(HttpStream stream, boolean hasBody) {
+                    response.hasBody = hasBody;
+                }
+
+                @Override
+                public int onResponseBody(HttpStream stream, ByteBuffer bodyBytesIn) {
+                    int start = bodyBytesIn.position();
+                    response.bodyBuffer.put(bodyBytesIn);
+                    int amountRead = bodyBytesIn.position() - start;
+
+                    // Slide the window open by the number of bytes just read
+                    return amountRead;
+                }
+
+                @Override
+                public void onResponseComplete(HttpStream stream, int errorCode) {
+                    response.onCompleteErrorCode = errorCode;
+                    reqCompleted.complete(null);
+                }
+
+                @Override
+                public boolean sendRequestBody(HttpStream stream, ByteBuffer bodyBytesOut) {
+                    transferData(bodyBytesIn, bodyBytesOut);
+
+                    return bodyBytesIn.remaining() == 0;
+                }
+            };
+
+            stream = conn.executeRequest(request, responseHandler);
+            Assert.assertNotNull(stream);
+            // Give the request up to 30 seconds to complete, otherwise throw a TimeoutException
+            reqCompleted.get(30, TimeUnit.SECONDS);
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        } finally {
+            if (stream != null) {
+                stream.close();
+            }
+            if (conn != null) {
+                conn.close();
+            }
+
+            /**
+             * TODO: This sleep is currently required to avoid a race condition between the conn.close() call above and
+             * the tlsContext.close() call below. Remove this Sleep when the bug is fixed in aws-c-io.
+             *
+             * The conn.close() call above schedules a shutdown task on the Native EventLoop. That task will shut down
+             * the s2n TLS connection and sends a TLS CloseNotify message to the TLS peer, and needs access to the
+             * TlsContext below.
+             *
+             * The tlsContext.close() call below immediately frees the native TlsContext structures used by s2n when
+             * sending the TLS CloseNotify message.
+             *
+             * Need to sleep here to allow enough time to pass for the task on the EventLoop to complete.
+             */
+            Thread.sleep(100);
```

Comment:
You need to wait until the connection finishes shutting down, like: `conn.shutdown().get()`

There might still be a bug where that event fires sliiiightly too early, but that's how the code ought to go.

Reply:
Updated tests to always wait for shutdown before closing. 

Do we want to update the [documentation here?](https://github.com/awslabs/aws-c-http/blob/master/include/aws/http/connection.h#L130-L137)

Something along the lines of this?
```
Calling release before shutdown has completed may cause race conditions and/or
Segfaults if you later free memory that can be used by shutdown tasks for this connection.
You should always wait for shutdown to complete before releasing.
```